stablehlo/dialect/TypeInference.h

```diff
@@ -0,0 +1,150 @@
+
+unsigned potentiallyComplexBitwidth(Type type);
+
+//===----------------------------------------------------------------------===//
+// Shape functions for ops.
+//===----------------------------------------------------------------------===//
+
+LogicalResult inferBatchNormGradOp(
+    Value operand, uint64_t featureIndex,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferBatchNormInferenceOp(
+    Value operand, SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferBatchNormTrainingOp(
+    Value operand, uint64_t featureIndex,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferCaseOp(
+    Optional<Location> location, RegionRange branches,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferDotGeneralOp(
+    Optional<Location> location, Value lhs, Value rhs,
+    ArrayRef<int64_t> lhsBatchingDims, ArrayRef<int64_t> rhsBatchingDims,
+    ArrayRef<int64_t> lhsContractingDims, ArrayRef<int64_t> rhsContractingDims,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferIfOp(
+    Optional<Location> location, RegionRange branches,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferMapOp(
+    Optional<Location> location, ValueRange inputs,
+    DenseIntElementsAttr dimensions, Region& computation,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+// TODO(https://github.com/openxla/stablehlo/issues/270)
+// Remove the util functions below when all the shape functions are moved here. 
+LogicalResult verifyReducerShape(
+    Optional<Location> loc, Block& block, ArrayRef<TensorType> inputArgTypes,
+    ArrayRef<TensorType> initValueTypes, int64_t numInputs,
+    ArrayRef<int64_t> allowedDimensions, bool allInputsUnranked,
+    SmallVectorImpl<TensorType>& accumulatorSubShapes);
+
+LogicalResult inferReduceOp(
+    Optional<Location> location, ValueRange inputs, ValueRange initValues,
+    DenseIntElementsAttr dimensions, Region& body,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferReduceWindowOp(
+    Optional<Location> location, ValueRange inputs, ValueRange initValues,
+    DenseIntElementsAttr windowDimensions,
+    Optional<DenseIntElementsAttr> windowStrides,
+    Optional<DenseIntElementsAttr> baseDilations,
+    Optional<DenseIntElementsAttr> windowDilations,
+    Optional<DenseIntElementsAttr> padding, Region& body,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferSortOp(
+    Optional<Location> location, ValueRange inputs, uint64_t dimension,
+    Region& comparator,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferTriangularSolveOp(
+    Optional<Location> location, Value a, Value b, bool leftSide,
+    bool is_transpose_a_invalid,
```

Comment:
This is a reasonable workaround for this particular op, but I imagine that for some other ops we might end up needing to declare enums right in this file.